kylin-spark-project/kylin-spark-engine/src/main/scala/org/apache/kylin/engine/spark/job/CuboidStatisticsJob.scala

```diff
@@ -0,0 +1,172 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kylin.engine.spark.job
+
+
+import org.apache.kylin.engine.spark.metadata.SegmentInfo
+import org.apache.kylin.measure.hllc.HLLCounter
+import org.apache.kylin.shaded.com.google.common.hash.{HashFunction, Hashing}
+import org.apache.spark.sql.{Dataset, Row}
+
+import scala.collection.mutable
+
+/**
+ * Calculate HLLCounter for each cuboid, to serve Cube Planner (to calculate cost and benefit of each cuboid).
+ */
+object CuboidStatisticsJob {
+
+  /**
+   * @param inputDs Part of FlatTable which contains all normal dimensions
+   * @return Cuboid level statistics data
+   */
+  def statistics(inputDs: Dataset[Row], seg: SegmentInfo): Array[(Long, AggInfo)] = {
+
+    val rkc = seg.allColumns.count(c => c.rowKey)
+    // maybe we should use sample operation to reduce cost later
+    val res = inputDs.rdd
+      .mapPartitions(new CuboidStatisticsJob(seg.getAllLayout.map(x => x.getId), rkc).statisticsWithinPartition)
+    val l = res.map(a => (a.key, a)).reduceByKey((a, b) => a.merge(b)).collect()
+    //    l.foreach(x => println(x._1 + " >>><<< " + x._2.cuboid.counter.getCountEstimate))
+    l
+  }
+}
+
+class CuboidStatisticsJob(ids: List[Long], rkc: Int) extends Serializable {
+  private val info = mutable.Map[Long, AggInfo]()
+  private var allCuboidsBitSet: Array[Array[Integer]] = Array()
+  private val hf: HashFunction = Hashing.murmur3_128
+  private val rowHashCodesLong = new Array[Long](rkc)
+  private var idx = 0
+  private var meter1 = 0L
+  private var meter2 = 0L
+  private var startMills = 0L
+  private var endMills = 0L
+
+
+  def statisticsWithinPartition(rows: Iterator[Row]): Iterator[AggInfo] = {
+    init()
+    println("CuboidStatisticsJob-statisticsWithinPartition1-" + System.currentTimeMillis())
+    rows.foreach(update)
+    printStat()
+    println("CuboidStatisticsJob-statisticsWithinPartition2-" + System.currentTimeMillis())
+    info.valuesIterator
+  }
+
+  def init(): Unit = {
+    println("CuboidStatisticsJob-Init1-" + System.currentTimeMillis())
+    allCuboidsBitSet = getCuboidBitSet(ids, rkc)
+    ids.foreach(i => info.put(i, AggInfo(i)))
+    println("CuboidStatisticsJob-Init2-" + System.currentTimeMillis())
+  }
+
+  def update(r: Row): Unit = {
+    idx += 1
+    if (idx <= 5)
+      println(r)
+    updateCuboid(r)
+  }
+
+  def updateCuboid(r: Row): Unit = {
+    // generate hash for each row key column
+    startMills = System.currentTimeMillis()
+    var idx = 0
+    while (idx < rkc) {
+      val hc = hf.newHasher
+      var colValue = r.get(idx).toString
+      if (colValue == null) colValue = "0"
```

Comment:
If r.get(idx) == null, r.get(idx).toString will throw NPE, please change to : 
`var colValue = if (r.get(idx) == null) "0" else r.get(idx).toString`